X11R6/contrib/programs/xsession-1.1/xsession.man
.\" [R.Q. 18-12-93] [R.Q. 11-01-94] [R.Q. 18-01-94] [R.Q. & A.N. 27-01-94]
.\" [R.Q. 07-02-94] [A.N. 05-04-94] [R.Q. 07-04-94] [R.Q. & A.N. 08-04-94]
.\" [A.N. 29-05-94] [R.Q. & A.N. 06-06-94]
.TH XSESSION 1 "6 June 1994" "Version 1.1"
.SH "NAME"
xsession - session manager for the X Window System
.SH "SYNOPSIS"
\fBxsession\fP [\-\fItoolkitoption\fP ...] [\-console] [\-cpp] [\-defaultWM ...] [\-D...] [\-m4] [\-keep] [\-notify] [\-remote] [\-sounds] [\-split] [\-startup] [\-version] [\-help]
.SH "DESCRIPTION"
The \fIxsession\fP program is a session manager.  It is normally executed by
your ~/.xinitrc (or ~/.xsession) script and controls your X Window session.
As soon as it is started, \fIxsession\fP launches a window manager and some
applications of your choice.  At anytime during your session, you may
switch to another window manager or execute some other applications from the
\fIxsession\fP menus.  You may also lock your screen using an independent
locker application and redirect to the xsession window the messages usually
sent to /dev/console.
.PP
One useful feature of \fIxsession\fP is the preprocessing of window managers'
configuration files.  When a window manager is launched, its startup file can
be filtered through \fIm4\fP(1) or \fIcpp\fP(1).  Some macros are defined
(such as the host name, screen size and so on) that will help you customize
your menus and screen layout.  One advantage of this feature is that several
window managers can share the same startup file.  For example, you can (and we
do) use the same file for \fImwm\fP(1) (Motif), \fItwm\fP(1), \fIctwm\fP(1) and
\fIvtwm\fP(1).  Some lines are protected by a test on the ``WM_NAME'' macro,
but the other lines are common.  This saves a lot of time if you want to
maintain the same menu structure for the four of them.  You can also use
``include'' statements in your file.
.PP
The best way to execute \fIxsession\fP is to include it in your ~/.xinitrc (or
~/.xsession) file, as the last client.  The \fIxinit\fP(1) manual page says
that ``the last long-lived program started [...] should be left in the
foreground so that the script won't exit''.  You will find a similar statement
in the \fIxdm\fP(1) manual page.  This last program should be
\fIxsession\fP.  As a matter of fact, this may even be the only program
executed in your startup file, because \fIxsession\fP itself provides a way to
launch other applications.  When started, \fIxsession\fP will look in your
home directory and execute a file named .X11Startup, if found.  This shell
script is used to start your client programs.
.SH "OPTIONS"
\fIXSession\fP accepts all of the standard X Toolkit command line options
(such as \fB\-iconic\fP, \fB\-geometry\fP and so on; see \fIX\fP(1)).  The
following additional arguments are supported.  Note that because the standard
X Toolkit method is used to retrieve X resources, most of the command line
options are only another way to set the value of some resources.  These
options always take precedence over their corresponding resources.
.TP 4
.B \-console
When \-console is set, xsession tries to redirect the console output messages
to its window.  This feature should work on most systems, but it may be
necessary that xsession is ``setuid'' to root.
.TP 4
.B \-defaultWM \fIwindow-manager's-short-name\fP
Specifies the (short) name of the default window manager to launch.  See the
``Menu customization'' section below for details.  The easiest way to choose
your default window manager is by using X resources in your ~/.Xdefaults file
(or the file you read with \fIxrdb\fP(1)).  The default value is ``TWM''.
.TP 4
.B \-keep
Sets the ``keepTmpFile'' resource to ``true''.  This option is used mainly
for debugging purposes.
.TP 4
.B \-notify, \-nonotify
When a message is printed in the \fIxsession\fP window and \-notify is set,
the icon name of the application has `` *'' appended, so that it is evident
even when the application is iconified.  When the application is de-iconified,
the icon name will be restored to its initial value.  The \-notify option is
set by default.
.TP 4
.B \-preprocessor \fIm4\fP|\fIcpp\fP
Specifies which preprocessor must be used.  Only two values may be used: either
``m4'' or ``cpp'' according as you want to use \fIm4\fP(1) or \fIcpp\fP(1).  
The default value is ``m4''.
.TP 4
.B \-quiet
If you use this option, \fIxsession\fP will not print any messages when an
application or window manager is started.  The warning messages
(e.g. ``configuration file not found'') will not be displayed either.
.TP 4
.B \-remote
You should set this option if you run \fIxsession\fP on a remote host (e.g. if
your ~/.xinitrc or ~/.xsession file tests if you are on a oldish machine and,
if so, executes \fIxsession\fP on a more powerful host in a remote shell).
\fIXSession\fP must then execute a special exit code to work around an unwanted
feature in \fIrsh\fP(1) that prevents it from exiting if there is one or more
children of \fIxsession\fP alive.  This exit code shuts down the socket
associated with the remote connection.
.TP 4
.B \-sounds, \-nosounds
Activates (or disables) the sounds.  The sounds will only be available if
\fIxsession\fP was compiled with SOUNDS defined.  The sounds are activated by
default if SOUNDS was defined, they are disabled otherwise.
.TP 4
.B \-split, \-nosplit
When \-split and \-console are set, the output messages of the programs and
the ones related to the console are shown in separate windows.  The \-split
option is set by default, but will be silently ignored if the ``console''
option is not used.
.TP 4
.B \-startup, \-nostartup
If \-startup is set, then the user's .X11Startup file is executed by
\fIxsession\fP at startup.  You may want to turn this option off for debugging
purposes or if you don't have a .X11Startup file.  The \-startup option is set
by default.
.PP
Note: if you use a ``\+'' instead of a ``\-'' in the boolean options above,
then the effect will be reversed (e.g. \fB\+startup\fP is the same as
\fB\-nostartup\fP).
.PP
There is no X resource equivalent for the following command line options:
.TP 4
.B \-D\fIsymbol\fP
This defines a symbol for the preprocessor that is run on the window managers'
configuration files.  All instances of \fIsymbol\fP in the configuration
files will be replaced by the NULL string (if \fIxsession\fP was compiled with
USE_M4 defined) or 1 (if \fIxsession\fP was compiled with USE_CPP).
.TP 4
.B \-D\fIsymbol\fP=\fIvalue\fP
This defines a symbol for the preprocessor that is run on the window managers'
configuration files.  All instances of \fIsymbol\fP in the configuration
files will be replaced by \fIvalue\fP.
.TP 4
.B \-cpp, \-m4
These options are equivalent to \-preprocessor cpp and \-preprocessor m4,
respectively.
.TP 4
.B \-help
Guess what?
.TP 4
.B \-version
Prints the current version number and exits.
.SH "X DEFAULTS"
This program uses mainly the Text, Paned, Box, MenuButton and SimpleMenu
\fIAthena\fP widgets.  Look in the \fIAthena Widget Set\fP documentation for
controlling them.  See also the ``Application resources'' section in this
manual page.
.SH "WIDGETS"
In order to specify resources, it is useful to know the hierarchy of
the widgets which compose \fIxsession\fP.  In the notation below,
indentation indicates hierarchical structure.  The widget class name
is given first, followed by the widget instance name.
.sp 
.nf
	XSession  xsession
		Paned  top_pane
			Box  menu_box
				MenuButton  session_button
					SimpleMenu  menu
						SmeBSB  lock_menu_item
						SmeBSB  quit_menu_item
				MenuButton  wm_button
					SimpleMenu  menu
						SmeBSB  <window manager's short name>_menu_item
						.
						. (one for each window manager)
						.
				MenuButton  app_button
					SimpleMenu  menu
						SmeBSB  <application's short name>_menu_item
						.
						. (one for each application)
						.
				MenuButton  misc_button
					SimpleMenu  menu
						SmeBSB  clear_menu_item
						SmeBSB  clear2_menu_item
			Text  ascii_text
			Text  ascii_text2
.fi
.sp 
The <window manager's short name> and <application's short name> represent the
short name of the program (a window manager or another application).  For
example, if you have ``TWM'' in your ``Window Managers'' menu, you will have a
SmeBSB instance named  ``TWM_menu_item''.
.SH "APPLICATION RESOURCES"
\fIXSession\fP has the following application-specific resources which allow
customizations unique to \fIxsession\fP.  Some of these resources have command
line equivalents which may be used to override them.
.PP
Any of these options may also be specified on the command line by using the
X Toolkit Intrinsics resource specification mechanism.  For example, to run
\fIxsession\fP using \fIxlock\fP(1) to lock the screen,
.nf
	% xsession \-xrm '*lockCommand:xlock -mode random'
.fi
.PP
.TP 8
\fBallowBell\fP (Class \fBAllowBell\fP)
Either ``true'' or ``false''.  When ``true'', this resource directs
\fIxsession\fP to ring the bell if the message printed in its window contains
one or more bell character(s).  This may happen if the output from one
application launched from the menu contains such characters (e.g. in an error
message).  Default value: ``false''.
.TP 8
\fBappNames\fP (Class \fBAppNames\fP)
This resource is a comma-separated list which gives the (short) names of the
applications to be shown in the ``Applications'' menu.  See the ``Menu
customization'' section below for details about names and short names of the
applications.  The default value should have been customized for your site
during the installation of \fIxsession\fP.  The names inside this resource are
used as \fI<application_name>\fP to build the following resource names:
.TP 8
\fBapp_\fP\fI<application_name>\fP (Class \fBApp_\fP\fI<application_name>\fP)
Each of these resources is a string that gives the full application name,
followed by a comma and the command to execute the application.  See the ``Menu
Customization'' section below for details about building these resources.
.TP 8
\fBconsole\fP (Class \fBConsole\fP)
Either ``true'' or ``false''.  If this resource is ``true'', xsession tries to
redirect the console output messages to its window.  This feature should work
on most systems, but it may be necessary that xsession is ``setuid'' to root.
Default value: ``false''.
.TP 8
\fBdefaultWM\fP (Class \fBDefaultWM\fP)
This resource contains the short name of the default window manager to launch.
See the ``Menu customization'' section below for details.  Default
value: ``TWM''.
.TP 8
\fBkeepTmpFile\fP (Class \fBKeepTmpFile\fP)
Either ``true'' or ``false''.  If this resource is ``true'', \fIxsession\fP
will not delete the temporary files created for the preprocessor and the
window manager.  This only makes sense if you use a preprocessor (such as
\fIm4\fP(1) or \fIcpp\fP(1)) to filter the window manager's configuration
file.  This option is used mainly for debugging purposes.  Default
value: ``false''.
.TP 8
\fBlockCommand\fP (Class \fBLockCommand\fP)
This resource contains the command to execute when the user selects the
``Lock'' entry in the ``Session'' menu.  The string can contain space
characters in order to add some arguments to the command.  Default value:
``xscreensaver-command -lock''.
.TP 8
\fBmotd\fP (Class \fBMotd\fP)
Either ``true'' or ``false''.  When this resource is ``true'',
the content of the /etc/motd file is printed in the \fIxsession\fP window at
startup.  Default value: ``true''.
.TP 8
\fBnotify\fP (Class \fBNotify\fP)
Either ``true'' or ``false''.  When a message is printed in the \fIxsession\fP
window and this resource is ``true'', the icon name of the
application has `` *'' appended, so that it is evident even when the
application is iconified.  When the application is de-iconified, the icon name
will be restored to its initial value.  Default value: ``true''.
.TP 8
\fBpreprocessor\fP (Class \fBPreprocessor\fP)
Specifies which preprocessor must be used.  Only two values may be used, either
``m4'' or ``cpp'', according as you want to use \fIm4\fP(1) or \fIcpp\fP(1).  
Default value: ``m4''.
.TP 8
\fBquiet\fP (Class \fBQuiet\fP)
Either ``true'' or ``false''.  If this resource is ``true'', \fIxsession\fP
will not print any messages when an application or window manager is started.
The warning messages (e.g. ``configuration file not found'') will not be
displayed either.  Default value: ``false''.
.TP 8
\fBremote\fP (Class \fBRemote\fP)
Either ``true'' or ``false''.  The usual way to change this value is by using a
command line option, not an X resource.  Default value: ``false''.
.TP 8
\fBringBell\fP (Class \fBRingBell\fP)
Either ``true'' or ``false''.  When a message is printed in the \fIxsession\fP
window and this resource is ``true'', the bell is rung so that the
user is warned even if the window is not visible.  Default value: ``false''.
.TP 8
\fBsounds\fP (Class \fBSounds\fP)
Either ``true'' or ``false''.  If this resource is ``false'', then the sounds
are disabled.  The sounds will only be available if \fIxsession\fP was
compiled with SOUNDS defined.  Default value: ``true'' if SOUNDS was defined,
else ``false''.
.TP 8
\fBsplitWindow\fP (Class \fBSplitWindow\fP)
Either ``true'' or ``false''.  If both this resource and the \fBconsole\fP
resource are ``true'', the normal output messages of the programs and the ones
related to the console are shown in separate windows.  Default value:
``true''.
.TP 8
\fBstartup\fP (Class \fBStartup\fP)
Either ``true'' or ``false''.  If this resource is ``true'', then
the user's .X11Startup file is executed by \fIxsession\fP at startup.  You may
want to turn this option off for debugging purposes or if you don't have
a .X11Startup file.  Default value: ``true''.
.TP 8

X11R6/contrib/programs/xsession-1.1/xsession.c
/* 
 * Copyright (C) 1993, 1994, Alain Nissen & Raphael Quinet,
 *                                                University of Liege, Belgium.
 *
 * The X Consortium, and any party obtaining a copy of these files from the
 * X Consortium, directly or indirectly, is granted, free of charge, a full
 * and unrestricted irrevocable, world-wide, paid up, royalty-free,
 * nonexclusive right and license to deal in this software and documentation
 * files (the "Software"), including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons who receive copies from any such party
 * to do so.  This license includes without limitation a license to do the
 * foregoing actions under any patents of the party supplying this software
 * to the X Consortium.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
 * X CONSORTIUM, THE AUTHORS OR THE UNIVERSITY OF LIEGE BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
 * USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * Authors :
 *    Raphael Quinet <quinet@montefiore.ulg.ac.be>
 *                   <quinet@stud.montefiore.ulg.ac.be>
 *    Alain Nissen   <nissen@montefiore.ulg.ac.be>
 *                   <nissen@stud.montefiore.ulg.ac.be>
 */

#if defined(_IBMR2) && defined(_AIX32)
#ifndef _POSIX_SOURCE
#define _POSIX_SOURCE
#endif
#endif

#include <X11/Xos.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xatom.h>
#include <X11/Xaw/Cardinals.h>
#include <X11/Xaw/AsciiText.h>
#include <X11/Xmu/Xmu.h>
#include <X11/Xmu/SysUtil.h>
#include <X11/Xmu/Atoms.h>
#include <X11/Xmu/StdSel.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include "xsession.h"
#include "resources.h"

#ifdef SOUNDS
#include <rplay.h>
#endif /* SOUNDS */

#if __STDC__
extern void input_ready_callback(Widget w, int *input, XtInputId *input_id);
extern void clear_proc(Widget w, XEvent *event, String params[],
                       Cardinal *num_params);
extern void clear2_proc(Widget w, XEvent *event, String params[],
                        Cardinal *num_params);
extern void iconify_proc(Widget w, XEvent *event, String params[],
                         Cardinal *num_params);
extern void deiconify_proc(Widget w, XEvent *event, String params[],
                           Cardinal *num_params);
extern void start_choosing_proc(Widget w, XEvent *event, String params[],
                                Cardinal *num_params);
extern void end_choosing_proc(Widget w, XEvent *event, String params[],
                              Cardinal *num_params);
extern void console_log(char *s);
extern void text_insert(Widget w, char *s, int len);
extern long text_length(Widget w);
extern char *_XawTextGetSTRING(TextWidget ctx, XawTextPosition left,
			       XawTextPosition right);
void wm_protocols_proc(Widget w, XEvent *event, String params[],
		       Cardinal *num_params);
#else
extern void input_ready_callback();
extern void clear_proc();
extern void clear2_proc();
extern void iconify_proc();
extern void deiconify_proc();
extern void start_choosing_proc();
extern void end_choosing_proc();
extern void console_log();
extern void text_insert();
extern long text_length();
extern char *_XawTextGetSTRING();
void wm_protocols_proc();
#endif

extern void activate_timeout();
extern void startup();
extern void print_motd();
extern void create_widgets();
extern void get_app_resources();
extern void get_wm_resources();
extern void end_current_wm();
extern void grimreaper();
extern int  redir_console();

extern int           errno;
extern Widget        top_shell, ascii_text, ascii_text2, clear2_menu_item;
extern pid_t         osm_child_pid;
extern Boolean       child_died;

char                *program_name;
int                  pipe_fd[2];
XtAppContext         app_context;
resources_rec        resources;
Boolean              widget_ok, closing_top_shell, suid;
XtInputId            console_id;

static char          mit_console_name[268] = "MIT_CONSOLE_";
static Atom          XA_WM_DELETE_WINDOW, XA_MIT_CONSOLE;
static XtInputId     input_id;
static int           return_code;
static XtActionsRec  actions[] =
  {
    {
      "WMProtocols",   wm_protocols_proc
    },
    {
      "Clear",         clear_proc
    },
    {
      "Clear2",        clear2_proc
    },
    {
      "Iconify",       iconify_proc
    },
    {
      "Deiconify",     deiconify_proc
    },
    {
      "StartChoosing", start_choosing_proc
    },
    {
      "EndChoosing",   end_choosing_proc
    }
  };


#ifdef SOUNDS
/*****************************************************************************
 *                                 play_sound                                *
 *---------------------------------------------------------------------------*
 * Plays a sound on the current display, using the rplay library.            *
 *****************************************************************************/

#if __STDC__
void play_sound(char *sound_name)
#else
void play_sound(sound_name)
     char *sound_name;
#endif
{
  if (resources.sounds && (sound_name != NULL) && strcmp(sound_name, "none"))
    {
      if (rplay_display(sound_name) < 0)
	console_log("Error: can't play a sound\n");
    }
}
#endif /* SOUNDS */


/*****************************************************************************
 *                               quit_callback                               *
 *---------------------------------------------------------------------------*
 * Terminates the program.                                                   *
 *****************************************************************************/

#if __STDC__
void quit_callback(Widget w, XtPointer client_data, XtPointer call_data)
#else
void quit_callback(w, client_data, call_data)
     Widget    w;
     XtPointer client_data, call_data;
#endif
{
  end_current_wm();
  if (osm_child_pid)
    {
      kill(osm_child_pid, SIGTERM);
      while (osm_child_pid)
	{
	  while (child_died == True)
	    grimreaper();
	  sleep(1);
	}
    }
  widget_ok = False;
  if (input_id)
    XtRemoveInput(input_id);
  if (console_id)
    XtRemoveInput(console_id);
  if (top_shell)
    XtDestroyWidget(top_shell);
  /*
   * The following hack shuts down the socket used by "rsh".  If xsession is
   * launched with a remote shell (e.g. "xon numbercruncher xsession -remote"),
   * this is the only way to be sure that the "rsh" ends as soon as you quit
   * xsession.  If you don't use the "-remote" option, the socket will remain
   * open as long as there is a child of xsession alive (e.g. xterm, emacs,...)
   * and the "rsh" won't end.
   */
  if (resources.remote)
    shutdown(fileno(stdout), 2);
#ifdef SOUNDS
  play_sound(resources.sound_end);
#endif /* SOUNDS */
  exit(return_code);
}


/*****************************************************************************
 *                                fatal_error                                *
 *---------------------------------------------------------------------------*
 * Prints an error message and terminates the program.                       *
 *****************************************************************************/

#if __STDC__
void fatal_error(int error_code, char *message)
#else
void fatal_error(error_code, message)
     int   error_code;
     char *message;
#endif
{
  return_code = error_code;
  fprintf(stderr, "%s: ", program_name);
  switch (error_code)
    {
      case ERR_SYNTAX :
        fprintf(stderr, "Syntax error.  Parameter \"%s\" not recognized.\n",
		message);
        break;
      case ERR_BAD_LIST :
        fprintf(stderr, "Invalid list.\n");
        break;
      case ERR_MEMORY :
        fprintf(stderr, "Out of memory!  Cannot allocate %u bytes\n",
		*((unsigned *) message));
        break;
      case ERR_PREPROCESSOR : 
	fprintf(stderr, "Syntax error.  Bad value for the `-preprocessor' option.\n");
	break;
      case ERR_SUID :
	fprintf(stderr, "Compilation error.  Setuid has not been enabled at compilation-time.\n");
	break;
      case ERR_ROOT :
	fprintf(stderr, "Setuid error.  Non root-setuid program.\n");
	break;
      default :
        fprintf(stderr, "Error code #%d.\n", error_code);
    }
  quit_callback(top_shell, (XtPointer) NULL, (XtPointer) NULL);
}


/*****************************************************************************
 *                                  zmalloc                                  *
 *---------------------------------------------------------------------------*
 * Just a wrapper around malloc().                                           *
 *****************************************************************************/

#if __STDC__
char *zmalloc(unsigned size)
#else
char *zmalloc(size)
     unsigned size;
#endif
{
  char *p;

  p = (char *) malloc(size);
  if (p == NULL)
    fatal_error(ERR_MEMORY, (char *)&size);
  return p;
}


/*****************************************************************************
 *                                  zrealloc                                 *
 *---------------------------------------------------------------------------*
 * Just a wrapper around realloc().                                          *
 *****************************************************************************/

#if __STDC__
char *zrealloc(char *ptr, unsigned size)
#else
char *zrealloc(ptr, size)
     char *ptr;
     unsigned size;
#endif
{
  char *p;

  p = (char *) realloc(ptr, size);
  if (p == NULL)
    fatal_error(ERR_MEMORY, (char *)&size);
  return p;
}


/*****************************************************************************
 *                               wm_protocols_proc                           *
 *---------------------------------------------------------------------------*
 * Handles WM_PROTOCOL client messages.  If a WM_DELETE_WINDOW is sent, then *
 * the program must exit.                                                    *
 *****************************************************************************/

#if __STDC__
void wm_protocols_proc(Widget w, XEvent *event, String params[],
		       Cardinal *num_params)
#else
void wm_protocols_proc(w, event, params, num_params)
     Widget    w;
     XEvent   *event;
     String    params[];
     Cardinal *num_params;
#endif
{
  if (event -> xclient.data.l[0] == XA_WM_DELETE_WINDOW)
    if (! closing_top_shell)
      {
        closing_top_shell = True;
        quit_callback(top_shell, (XtPointer) NULL, (XtPointer) NULL);
        closing_top_shell = False;
      }
}


/* The three following functions (selection_lost, convert_selection and      */
/* copy_selection_callback) were taken and munged from "xconsole.c".         */
/*                                        XConsole is (C) 1990 X Consortium. */

/*****************************************************************************
 *                               selection_lost                              *
 *---------------------------------------------------------------------------*
 * Warns the user that the console redirection is now inactive.              *
 *****************************************************************************/

#if __STDC__
static void selection_lost(Widget w, Atom *selection)
#else
static void selection_lost(w, selection)
     Widget  w;
     Atom   *selection;
#endif
{
  if (!resources.quiet)
    console_log("Warning: console output lost.\n");
  if (console_id)
    {
      XtRemoveInput(console_id);
      console_id = 0;
    }
  if (resources.split_window)
    {
      Arg arg[1];

      resources.split_window = False;
      XtUnmanageChild(ascii_text2);
      XtSetArg(arg[0], XtNsensitive, False);
      XtSetValues(clear2_menu_item, arg, ONE);
    }
}


/*****************************************************************************
 *                             convert_selection                             *
 *---------------------------------------------------------------------------*
 * Is called whenever someone requests the current value of the selection;   *
 * we can offer the console output only if separate windows are used.        *
 *****************************************************************************/

#if __STDC__
static Boolean convert_selection(Widget w, Atom *selection, Atom *target,
				 Atom *type, XtPointer *value,
				 unsigned long *length, int *format)
#else
static Boolean convert_selection(w, selection, target, type, value, length,
				 format)
     Widget         w;
     Atom          *selection, *target, *type;
     XtPointer     *value;
     unsigned long *length;
     int           *format;
#endif
{
  Display                *d;
  XSelectionRequestEvent *req;
  Atom                   *targetP, *std_targets;
  unsigned long           std_length;
  long                   *temp;

  if (resources.console && resources.split_window)
    {
      d = XtDisplay(w);
      req = XtGetSelectionRequest(w, *selection, (XtRequestId) NULL);
      if (*target == XA_TARGETS(d))
	{
	  XmuConvertStandardSelection(w, req->time, selection, target, type,
				      (caddr_t *) &std_targets, &std_length,
				      format);
	  *value = (XtPointer) XtMalloc(sizeof(Atom) * (std_length + 5));
	  targetP = *(Atom **) value;
	  *targetP++ = XA_STRING;
	  *targetP++ = XA_TEXT(d);
	  *targetP++ = XA_LENGTH(d);
	  *targetP++ = XA_LIST_LENGTH(d);
	  *targetP++ = XA_CHARACTER_POSITION(d);
	  *length = std_length + (targetP - (*(Atom **) value));
	  memcpy((char *) std_targets, (char *) targetP, sizeof(Atom) *
		 std_length);
	  XtFree((char *) std_targets);
	  *type = XA_ATOM;
	  *format = 32;
	  return True;
	}
      if (*target == XA_LIST_LENGTH(d) || *target == XA_LENGTH(d))
	{
	  temp = (long *) XtMalloc(sizeof(long));
	  if (*target == XA_LIST_LENGTH(d))
	    *temp = 1L;
	  else			/* *target == XA_LENGTH(d) */
	    *temp = (long) text_length(ascii_text2);
	  *value = (XtPointer) temp;
	  *type = XA_INTEGER;
	  *length = 1L;
	  *format = 32;
	  return True;
	}
      if (*target == XA_CHARACTER_POSITION(d))
	{
	  temp = (long *) XtMalloc(2 * sizeof(long));
	  temp[0] = (long) 0;
	  temp[1] = text_length(ascii_text2);
	  *value = (XtPointer) temp;
	  *type = XA_SPAN(d);
	  *length = 2L;
	  *format = 32;
	  return True;
	}
      if (*target == XA_STRING || *target == XA_TEXT(d) ||
	  *target == XA_COMPOUND_TEXT(d))
	{
	  if (*target == XA_COMPOUND_TEXT(d))
	    *type = *target;
	  else
	    *type = XA_STRING;
	  *length = text_length(ascii_text2);
	  *value = (XtPointer) _XawTextGetSTRING((TextWidget) ascii_text2, 0,
						 *length);
	  *format = 8;
	  return True;
	}
      if (XmuConvertStandardSelection(w, req->time, selection, target, type,
				      (caddr_t *) value, length, format))
	return True;
      return False;
    }
  else
    return False;
}


/*****************************************************************************
 *                          copy_selection_callback                          *
 *---------------------------------------------------------------------------*
 * Copies the content of the running XConsole window to our own window.      *
 *****************************************************************************/

#if __STDC__
static void copy_selection_callback(Widget w, XtPointer client_data,
				    Atom *selection, Atom *type,
				    XtPointer value, unsigned long *length,
				    int *format)
#else
static void copy_selection_callback(w, client_data, selection, type, value,
				    length, format)
     Widget         w;
     XtPointer      client_data;
     Atom          *selection, *type;
     XtPointer      value;
     unsigned long *length;
     int           *format;
#endif
{
  if (*type != XT_CONVERT_FAIL)
    text_insert(ascii_text, (char *) value, *length);
  XtOwnSelection(top_shell, XA_MIT_CONSOLE, CurrentTime, convert_selection,
		 selection_lost, NULL);
}
    

/*****************************************************************************
 *                                    main                                   *
 *---------------------------------------------------------------------------*
 * Believe me: the program won't run without this function...                *
 *****************************************************************************/

#if __STDC__
#if defined(sgi) || defined(__sgi) || defined(__sgi__)
void main(unsigned int argc, char *argv[])
#else
void main(int argc, char *argv[])
#endif
#else
void main(argc, argv)
#if defined(sgi) || defined(__sgi) || defined(__sgi__)
     unsigned int argc;
#else
     int   argc;
#endif
     char *argv[];
#endif
{
  char    *p;
  int      console;
  Boolean  suid_fail;

  if (getuid() != geteuid())
    {
      suid = True;
#ifndef ENABLE_SUID
      suid_fail = True;
#else
      suid_fail = False;
#endif /* ENABLE_SUID */
    }
  else
    {
      suid = False;
      suid_fail = False;
    }
  if ((program_name = strrchr(argv[0], (int) '/')) == NULL)
    program_name = argv[0];
  else
    program_name ++;
  if ((suid == True) && (suid_fail == True))
    fatal_error(ERR_SUID, (char *) NULL);
  if ((suid == True) && geteuid())
    fatal_error(ERR_ROOT, (char *) NULL);
  return_code = 0;
  widget_ok = False;
  top_shell = XtAppInitialize(&app_context, "XSession",
                              options, XtNumber(options),
			      &argc, (String *) argv,
                              fallback_resources, NULL, ZERO);
  XA_WM_DELETE_WINDOW = XInternAtom(XtDisplay(top_shell),
				    "WM_DELETE_WINDOW", False);
  XtGetApplicationResources(top_shell, (XtPointer) &resources,
                            resources_desc, XtNumber(resources_desc),
			    NULL, ZERO);
  XtAppAddActions(app_context, actions, XtNumber(actions));

  resources.num_defines = 0;
  for (argv++, argc--; argc; argv++, argc--)
    if (!strcmp(*argv, "-help") || !strcmp(*argv, "--help")
	|| !strcmp(*argv, "-usage") || !strcmp(*argv, "--usage"))
      {
	printf("Usage: %s [-toolkitoption ...] [-console] [-cpp] [-defaultWM <name>]\n",
	       program_name);
	printf("                [-D<symbol>[=<value>]] [-keep] [-m4] [-notify]\n");
#ifdef SOUNDS
	printf("                [-preprocessor m4|cpp] [-remote] [-sounds] [-split] [-startup]\n");
        printf("                [-version] [-help]\n");
#else
	printf("                [-preprocessor m4|cpp] [-remote] [-split] [-startup] [-version]\n");
        printf("                [-help]\n");
#endif /* SOUNDS */
        quit_callback(top_shell, (XtPointer) NULL, (XtPointer) NULL);
      }
    else
      if (!strcmp(*argv, "-version") || !strcmp(*argv, "--version"))
	{
	  printf("XSession %s\n", VERSION);
	  quit_callback(top_shell, (XtPointer) NULL, (XtPointer) NULL);
	}
      else
	if ((argv[0][0] == '-') && (argv[0][1] == 'D'))   /* defines */
	  {
	    p = *argv + 2;
	    while (*p && (*p != '='))
	      p++;
	    if (*p == '=')
	      {
		*p = '\0';
		p++;
	      }
	    if (resources.num_defines++ == 0)
	      {
		resources.defines_sym = (String *) zmalloc(sizeof(String));
		resources.defines_val = (String *) zmalloc(sizeof(String));
		resources.defines_sym[0] = *argv + 2;
		resources.defines_val[0] = p;
	      }
	    else
	      {
		resources.defines_sym =
		  (String *) zrealloc((char *) resources.defines_sym,
				      resources.num_defines * sizeof(String));
		resources.defines_val =
		  (String *) zrealloc((char *) resources.defines_val,
				      resources.num_defines * sizeof(String));
		resources.defines_sym[resources.num_defines - 1] = *argv + 2;
		resources.defines_val[resources.num_defines - 1] = p;
	      }
	  }
	else
	  fatal_error(ERR_SYNTAX, *argv);
  if (strcmp(resources.preprocessor, "cpp") && strcmp(resources.preprocessor, "m4"))
    fatal_error(ERR_PREPROCESSOR, (char *) NULL);
  get_wm_resources();
  get_app_resources();
  p = getenv("HOME");
  if (p != NULL)
    chdir(p);
  if (getenv("DISPLAY") == NULL)
    {
      p = (char *) zmalloc((10 + strlen(DisplayString(XtDisplay(top_shell))))
			   * sizeof(char));
      sprintf(p, "DISPLAY=%s", DisplayString(XtDisplay(top_shell)));
      putenv(p);
    }
  if (!resources.console)
    resources.split_window = False;
  create_widgets();
  XtRealizeWidget(top_shell);
  widget_ok = True;
  XSetWMProtocols(XtDisplay(top_shell), XtWindow(top_shell),
		  &XA_WM_DELETE_WINDOW, 1);
  pipe(pipe_fd);
  input_id = XtAppAddInput(app_context, pipe_fd[0],
			   (XtPointer) XtInputReadMask,
                           (XtInputCallbackProc) input_ready_callback, NULL);
  activate_timeout();
  if (resources.motd)
    print_motd();
  osm_child_pid = 0;
  if (resources.console)
    {
      XmuGetHostname(mit_console_name + 12, 255);
      XA_MIT_CONSOLE = XInternAtom(XtDisplay(top_shell), mit_console_name,
				   False);
      if (XGetSelectionOwner(XtDisplay(top_shell), XA_MIT_CONSOLE))
	XtGetSelectionValue(top_shell, XA_MIT_CONSOLE, XA_STRING,
			    copy_selection_callback, NULL, CurrentTime);
      else
	XtOwnSelection(top_shell, XA_MIT_CONSOLE, CurrentTime,
		       convert_selection, selection_lost, NULL);
      console = redir_console();
      if (console >= 0)
	{
	  console_id = XtAppAddInput(app_context, console,
				     (XtPointer) XtInputReadMask,
				     (XtInputCallbackProc)
				     input_ready_callback, NULL);
	}
      else
	{
	  console_id = 0;
	  if (!resources.quiet)
	    console_log("Warning: cannot redirect console messages.\n");
	  if (resources.split_window)
	    {
	      Arg arg[1];

	      resources.split_window = False;
	      XtUnmanageChild(ascii_text2);
	      XtSetArg(arg[0], XtNsensitive, False);
	      XtSetValues(clear2_menu_item, arg, ONE);
	    }
	}
    }
  else
    console_id = 0;
  if (suid == True)
    {
      setgid(getgid());
      setuid(getuid());
    }
#ifdef SOUNDS
  play_sound(resources.sound_start);
#endif /* SOUNDS */
  startup();
  XtAppMainLoop(app_context);
}
